File: sdk/samples/FrameworkSamples/CLR/GCDemo/cs/formmain.cs
//-----------------------------------------------------------------------
//  This file is part of the Microsoft .NET SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//-----------------------------------------------------------------------
using System;
using System.Globalization;
using System.Security;
using System.Security.Permissions;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Windows.Forms.Design;
using System.Data;
using System.Threading;
using System.Runtime.InteropServices;

[assembly: CLSCompliant(true)]
[assembly: ComVisible(false)]
[assembly: FileIOPermissionAttribute(SecurityAction.RequestMinimum)]

namespace Microsoft.Samples.GCDemo
{
    unsafe class formMain : System.Windows.Forms.Form
    {
        private int _allocation;
        private System.Windows.Forms.Label labelTotCount;
        private int _curCount;
        private System.Windows.Forms.Label labelCurCount;
        private System.Windows.Forms.Label labelTotal;
        private System.Windows.Forms.Button buttonClear;
        private int _totCount;

        private System.Windows.Forms.TreeView treeMain;
        private int _gcBaseLine;

        private System.Windows.Forms.Button buttonExit;
        private System.Windows.Forms.GroupBox groupPressure;
        private System.Windows.Forms.RadioButton radioPressureNo;
        private System.Windows.Forms.Button buttonRun;
        private System.Windows.Forms.RadioButton radioPressureYes;
        private System.Windows.Forms.Label labelAllocation;
        private System.Windows.Forms.GroupBox groupVariables;
        private System.Windows.Forms.Label label1;
        private System.Windows.Forms.ComboBox comboAmount;
        private System.Windows.Forms.Label label2;
        private System.Windows.Forms.NumericUpDown numericAllocations;
        private System.Windows.Forms.GroupBox groupCollections;
        private System.Windows.Forms.Label labelCount;
        private System.Windows.Forms.Label labelDesc;
        private System.ComponentModel.Container components = null;

        public formMain()
        {
            InitializeComponent();
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                if (components != null)
                {
                    components.Dispose();
                }
            }
            base.Dispose(disposing);
        }

        #region Windows Form Designer generated code
        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            System.ComponentModel.ComponentResourceManager resources = new System.ComponentModel.ComponentResourceManager(typeof(formMain));
            this.buttonExit = new System.Windows.Forms.Button();
            this.groupPressure = new System.Windows.Forms.GroupBox();
            this.radioPressureYes = new System.Windows.Forms.RadioButton();
            this.radioPressureNo = new System.Windows.Forms.RadioButton();
            this.buttonRun = new System.Windows.Forms.Button();
            this.labelAllocation = new System.Windows.Forms.Label();
            this.groupVariables = new System.Windows.Forms.GroupBox();
            this.numericAllocations = new System.Windows.Forms.NumericUpDown();
            this.label2 = new System.Windows.Forms.Label();
            this.comboAmount = new System.Windows.Forms.ComboBox();
            this.label1 = new System.Windows.Forms.Label();
            this.groupCollections = new System.Windows.Forms.GroupBox();
            this.labelTotal = new System.Windows.Forms.Label();
            this.labelCurCount = new System.Windows.Forms.Label();
            this.labelTotCount = new System.Windows.Forms.Label();
            this.labelCount = new System.Windows.Forms.Label();
            this.labelDesc = new System.Windows.Forms.Label();
            this.buttonClear = new System.Windows.Forms.Button();
            this.treeMain = new System.Windows.Forms.TreeView();
            this.groupPressure.SuspendLayout();
            this.groupVariables.SuspendLayout();
            ((System.ComponentModel.ISupportInitialize)(this.numericAllocations)).BeginInit();
            this.groupCollections.SuspendLayout();
            this.SuspendLayout();
            // 
            // buttonExit
            // 
            resources.ApplyResources(this.buttonExit, "buttonExit");
            this.buttonExit.Name = "buttonExit";
            this.buttonExit.Click += new System.EventHandler(this.buttonExit_Click);
            // 
            // groupPressure
            // 
            this.groupPressure.Controls.Add(this.radioPressureYes);
            this.groupPressure.Controls.Add(this.radioPressureNo);
            resources.ApplyResources(this.groupPressure, "groupPressure");
            this.groupPressure.Name = "groupPressure";
            this.groupPressure.TabStop = false;
            // 
            // radioPressureYes
            // 
            resources.ApplyResources(this.radioPressureYes, "radioPressureYes");
            this.radioPressureYes.Name = "radioPressureYes";
            // 
            // radioPressureNo
            // 
            this.radioPressureNo.Checked = true;
            resources.ApplyResources(this.radioPressureNo, "radioPressureNo");
            this.radioPressureNo.Name = "radioPressureNo";
            // 
            // buttonRun
            // 
            resources.ApplyResources(this.buttonRun, "buttonRun");
            this.buttonRun.Name = "buttonRun";
            this.buttonRun.Click += new System.EventHandler(this.buttonRun_Click);
            // 
            // labelAllocation
            // 
            resources.ApplyResources(this.labelAllocation, "labelAllocation");
            this.labelAllocation.Name = "labelAllocation";
            // 
            // groupVariables
            // 
            this.groupVariables.Controls.Add(this.numericAllocations);
            this.groupVariables.Controls.Add(this.label2);
            this.groupVariables.Controls.Add(this.comboAmount);
            this.groupVariables.Controls.Add(this.label1);
            resources.ApplyResources(this.groupVariables, "groupVariables");
            this.groupVariables.Name = "groupVariables";
            this.groupVariables.TabStop = false;
            // 
            // numericAllocations
            // 
            resources.ApplyResources(this.numericAllocations, "numericAllocations");
            this.numericAllocations.Minimum = new decimal(new int[] {
            1,
            0,
            0,
            0});
            this.numericAllocations.Name = "numericAllocations";
            this.numericAllocations.Value = new decimal(new int[] {
            9,
            0,
            0,
            0});
            this.numericAllocations.ValueChanged += new System.EventHandler(this.numericAllocations_ValueChanged);
            // 
            // label2
            // 
            resources.ApplyResources(this.label2, "label2");
            this.label2.Name = "label2";
            // 
            // comboAmount
            // 
            this.comboAmount.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
            this.comboAmount.FormattingEnabled = true;
            resources.ApplyResources(this.comboAmount, "comboAmount");
            this.comboAmount.Name = "comboAmount";
            this.comboAmount.SelectedIndexChanged += new System.EventHandler(this.comboAmount_SelectedIndexChanged);
            // 
            // label1
            // 
            resources.ApplyResources(this.label1, "label1");
            this.label1.Name = "label1";
            // 
            // groupCollections
            // 
            this.groupCollections.Controls.Add(this.labelTotal);
            this.groupCollections.Controls.Add(this.labelCurCount);
            this.groupCollections.Controls.Add(this.labelTotCount);
            this.groupCollections.Controls.Add(this.labelCount);
            resources.ApplyResources(this.groupCollections, "groupCollections");
            this.groupCollections.Name = "groupCollections";
            this.groupCollections.TabStop = false;
            // 
            // labelTotal
            // 
            resources.ApplyResources(this.labelTotal, "labelTotal");
            this.labelTotal.Name = "labelTotal";
            // 
            // labelCurCount
            // 
            resources.ApplyResources(this.labelCurCount, "labelCurCount");
            this.labelCurCount.Name = "labelCurCount";
            // 
            // labelTotCount
            // 
            resources.ApplyResources(this.labelTotCount, "labelTotCount");
            this.labelTotCount.Name = "labelTotCount";
            // 
            // labelCount
            // 
            resources.ApplyResources(this.labelCount, "labelCount");
            this.labelCount.Name = "labelCount";
            // 
            // labelDesc
            // 
            resources.ApplyResources(this.labelDesc, "labelDesc");
            this.labelDesc.Name = "labelDesc";
            // 
            // buttonClear
            // 
            resources.ApplyResources(this.buttonClear, "buttonClear");
            this.buttonClear.Name = "buttonClear";
            this.buttonClear.Click += new System.EventHandler(this.buttonClear_Click);
            // 
            // treeMain
            // 
            resources.ApplyResources(this.treeMain, "treeMain");
            this.treeMain.Name = "treeMain";
            // 
            // formMain
            // 
            resources.ApplyResources(this, "$this");
            this.Controls.Add(this.treeMain);
            this.Controls.Add(this.buttonClear);
            this.Controls.Add(this.labelDesc);
            this.Controls.Add(this.groupCollections);
            this.Controls.Add(this.groupVariables);
            this.Controls.Add(this.labelAllocation);
            this.Controls.Add(this.buttonRun);
            this.Controls.Add(this.groupPressure);
            this.Controls.Add(this.buttonExit);
            this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog;
            this.MaximizeBox = false;
            this.Name = "formMain";
            this.Load += new System.EventHandler(this.formMain_Load);
            this.groupPressure.ResumeLayout(false);
            this.groupVariables.ResumeLayout(false);
            ((System.ComponentModel.ISupportInitialize)(this.numericAllocations)).EndInit();
            this.groupCollections.ResumeLayout(false);
            this.ResumeLayout(false);

        }
        #endregion

        [STAThread]
        static void Main()
        {
            Application.Run(new formMain());
        }

        private void formMain_Load(System.Object sender, System.EventArgs e)
        {
            _allocation = 1000;
            _curCount = 0;
            _totCount = 0;
            comboAmount.Items.Add("1000");
            comboAmount.Items.Add("5000");
            comboAmount.Items.Add("10000");
            comboAmount.Items.Add("100000");
            comboAmount.Items.Add("1000000");
            comboAmount.Items.Add("10000000");
            comboAmount.Items.Add("100000000");
            comboAmount.SelectedIndex = 0;
            UpdateDescLabel();
        }

        private void UpdateAllocationLabel()
        {
            labelAllocation.Text = String.Format(CultureInfo.InvariantCulture, "Allocate {0} bytes of available memory", comboAmount.Text);
        }

        private void UpdateDescLabel()
        {
            labelDesc.Text = String.Format(CultureInfo.InvariantCulture, "This demo allocates {0} bytes of available unmanaged memory, {1} times", comboAmount.Text, numericAllocations.Value);
        }

        private void buttonRun_Click(System.Object sender, System.EventArgs e)
        {
            Cursor = Cursors.WaitCursor;


            // Garbage collection is forced to "clean" the heap, then
            // we pause while waiting for the separate finalizer thread 
            // to clear the finalization queue. A final GC.Collect()
            // call clears out anything that may have been resurrected
            // by the finalizer.
            GC.Collect();
            GC.WaitForPendingFinalizers();
            GC.Collect();

            MEMORYSTATUSEX ms = new MEMORYSTATUSEX();

            ms.Init();
            NativeMethods.GlobalMemoryStatusEx(ref ms);

            int pressure = _allocation;

            treeMain.Nodes.Clear();

            for (int i = 0; i < numericAllocations.Value; i++)
            {
                int nodeElement = treeMain.Nodes.Add(new TreeNode(String.Format(CultureInfo.InvariantCulture, "Allocating 1/{0} for the {1}{2} time", _allocation, i + 1, GetSuffix(i))));

                MemoryHogger m = new MemoryHogger(pressure, radioPressureYes.Checked);
                TreeNode tn = new TreeNode();

                if (m.Success)
                {
                    tn.Text = "\tAllocation Succeeded!";
                    tn.BackColor = Color.FromArgb(0, 255, 0);
                    treeMain.Nodes[nodeElement].Nodes.Add(tn);
                }
                else
                {
                    tn.Text = "\tAllocation Failed!";
                    tn.BackColor = Color.Red;
                    treeMain.Nodes[nodeElement].Nodes.Add(tn);
                }
                tn.Parent.Expand();
            }
            _curCount = GC.CollectionCount(0) - _gcBaseLine - _totCount;
            _totCount = GC.CollectionCount(0) - _gcBaseLine;
            labelCount.Text = _curCount.ToString(CultureInfo.InvariantCulture);
            labelTotCount.Text = _totCount.ToString(CultureInfo.InvariantCulture);
            buttonClear.Enabled = true;
            Cursor = Cursors.Arrow;
        }

        private string GetSuffix(int i)
        {
            if (i == 0)
                return "st";
            else if (i == 1)
                return "nd";
            else if (i == 2)
                return "rd";
            else
                return "th";
        }

        private void buttonExit_Click(System.Object sender, System.EventArgs e)
        {
            this.Close();
        }

        private void comboAmount_SelectedIndexChanged(System.Object sender, System.EventArgs e)
        {
            _allocation = Convert.ToInt32(comboAmount.Text);
            UpdateDescLabel();
            UpdateAllocationLabel();
        }

        private void numericAllocations_ValueChanged(System.Object sender, System.EventArgs e)
        {
            UpdateDescLabel();
        }
        private void buttonClear_Click(object sender, System.EventArgs e)
        {
            treeMain.Nodes.Clear();
            labelCount.Text = "0";
            labelTotCount.Text = "0";
            _gcBaseLine += _totCount;
            _curCount = 0;
            _totCount = 0;

            buttonClear.Enabled = false;
        }

        class MemoryHogger : IDisposable
        {
            int pressure;
            bool addedPressure;
            bool allocated;
            IntPtr memptr;

            const int MEM_RESERVE = 0x2000;
            const int MEM_RELEASE = 0x8000;
            const int PAGE_READWRITE = 4;

            // Allocated unmanaged memory and, optionally, inform the
            // GC of the added memory
            public MemoryHogger(int size, bool addPressure)
            {
                if (addPressure)
                {
                    addedPressure = true;
                    pressure = size;
                    GC.AddMemoryPressure(pressure);
                }

                memptr = NativeMethods.VirtualAlloc((IntPtr)0, size, MEM_RESERVE, PAGE_READWRITE);
                allocated = (memptr != IntPtr.Zero);

                if (addedPressure && !allocated)
                {
                    GC.RemoveMemoryPressure(pressure);
                    addedPressure = false;
                }
            }

            // Finalization code for the MemoryHogger - make sure to reduce
            // the GC pressure by EXACTLY the amount that was added
            ~MemoryHogger()
            {
                if (allocated)
                {
                    NativeMethods.VirtualFree(memptr, 0, MEM_RELEASE);
                }

                if (addedPressure)
                {
                    GC.RemoveMemoryPressure(pressure);
                }
            }

            void IDisposable.Dispose() { }

            public bool Success
            {
                get { return allocated; }
            }
        }
    }
}

File: sdk/samples/FrameworkSamples/CLR/Stopwatch/vb/formmain.vb
'-----------------------------------------------------------------------
'  This file is part of the Microsoft .NET SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'-----------------------------------------------------------------------
Imports System.Threading
Imports System.Globalization

' This is a simple demo that demonstrates the new StopWatch API
' You'll notice in this demo that most of the code
' is simply to draw the form! The StopWatch code itself
' is a cinch
Namespace Microsoft.Samples.StopWatchDemo
    Public Class FormMain
        Inherits System.Windows.Forms.Form

#Region " Windows Form Designer generated code "

        Public Sub New()
            MyBase.New()

            'This call is required by the Windows Form Designer.
            InitializeComponent()

            'Add any initialization after the InitializeComponent() call

        End Sub

        'Form overrides dispose to clean up the component list.
        Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If Not (components Is Nothing) Then
                    components.Dispose()
                End If
            End If
            MyBase.Dispose(disposing)
        End Sub

        'Required by the Windows Form Designer
        Private components As System.ComponentModel.IContainer

        'NOTE: The following procedure is required by the Windows Form Designer
        'It can be modified using the Windows Form Designer.  
        'Do not modify it using the code editor.
        Private WithEvents labelExit As System.Windows.Forms.Label
        Private WithEvents labelTime As System.Windows.Forms.Label
        Private WithEvents timerMain As System.Windows.Forms.Timer
        Private WithEvents buttonLap As System.Windows.Forms.Button
        Private WithEvents buttonStart As System.Windows.Forms.Button
        Private WithEvents labelLapPrompt As System.Windows.Forms.Label
        Private WithEvents LabelTimePrompt As System.Windows.Forms.Label
        Private WithEvents labelLap As System.Windows.Forms.Label
        Private WithEvents labelBottomLeft As System.Windows.Forms.Label
        Private WithEvents labelBottomRight As System.Windows.Forms.Label
        Private WithEvents labelTopLeft As System.Windows.Forms.Label
        Private WithEvents labelTopRight As System.Windows.Forms.Label
        <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
            Me.labelExit = New System.Windows.Forms.Label
            Me.labelTime = New System.Windows.Forms.Label
            Me.buttonLap = New System.Windows.Forms.Button
            Me.components = New System.ComponentModel.Container
            Me.timerMain = New System.Windows.Forms.Timer(Me.components)
            Me.labelLap = New System.Windows.Forms.Label
            Me.buttonStart = New System.Windows.Forms.Button
            Me.labelLapPrompt = New System.Windows.Forms.Label
            Me.LabelTimePrompt = New System.Windows.Forms.Label
            Me.labelBottomLeft = New System.Windows.Forms.Label
            Me.labelBottomRight = New System.Windows.Forms.Label
            Me.labelTopLeft = New System.Windows.Forms.Label
            Me.labelTopRight = New System.Windows.Forms.Label
            Me.SuspendLayout()
            '
            'labelExit
            '
            Me.labelExit.BorderStyle = System.Windows.Forms.BorderStyle.Fixed3D
            Me.labelExit.Location = New System.Drawing.Point(88, 4)
            Me.labelExit.Margin = New System.Windows.Forms.Padding(3, 3, 3, 1)
            Me.labelExit.Name = "labelExit"
            Me.labelExit.Size = New System.Drawing.Size(20, 21)
            Me.labelExit.TabIndex = 0
            Me.labelExit.Text = "x"
            Me.labelExit.TextAlign = System.Drawing.ContentAlignment.MiddleCenter
            '
            'labelTime
            '
            Me.labelTime.Font = New System.Drawing.Font("Microsoft Sans Serif", 20.25!, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, CType(0, Byte))
            Me.labelTime.ForeColor = System.Drawing.Color.Yellow
            Me.labelTime.Location = New System.Drawing.Point(8, 52)
            Me.labelTime.Margin = New System.Windows.Forms.Padding(3, 3, 3, 1)
            Me.labelTime.Name = "labelTime"
            Me.labelTime.Size = New System.Drawing.Size(188, 33)
            Me.labelTime.TabIndex = 1
            Me.labelTime.Text = "00:00:00.00"
            '
            'buttonLap
            '
            Me.buttonLap.Font = New System.Drawing.Font("Comic Sans MS", 14.25!, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, CType(0, Byte))
            Me.buttonLap.ForeColor = System.Drawing.Color.Yellow
            Me.buttonLap.Location = New System.Drawing.Point(102, 154)
            Me.buttonLap.Margin = New System.Windows.Forms.Padding(2, 3, 1, 3)
            Me.buttonLap.Name = "buttonLap"
            Me.buttonLap.Size = New System.Drawing.Size(72, 34)
            Me.buttonLap.TabIndex = 4
            Me.buttonLap.Text = "Lap"
            '
            'timerMain
            '
            Me.timerMain.Enabled = True
            Me.timerMain.Interval = 50
            '
            'labelLap
            '
            Me.labelLap.Font = New System.Drawing.Font("Microsoft Sans Serif", 20.25!, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, CType(0, Byte))
            Me.labelLap.ForeColor = System.Drawing.Color.Yellow
            Me.labelLap.Location = New System.Drawing.Point(8, 112)
            Me.labelLap.Name = "labelLap"
            Me.labelLap.Size = New System.Drawing.Size(191, 34)
            Me.labelLap.TabIndex = 5
            Me.labelLap.Visible = False
            '
            'buttonStart
            '
            Me.buttonStart.FlatStyle = System.Windows.Forms.FlatStyle.Flat
            Me.buttonStart.Font = New System.Drawing.Font("Comic Sans MS", 14.25!, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, CType(0, Byte))
            Me.buttonStart.ForeColor = System.Drawing.Color.Yellow
            Me.buttonStart.Location = New System.Drawing.Point(26, 154)
            Me.buttonStart.Margin = New System.Windows.Forms.Padding(1, 3, 2, 3)
            Me.buttonStart.Name = "buttonStart"
            Me.buttonStart.Size = New System.Drawing.Size(72, 34)
            Me.buttonStart.TabIndex = 2
            Me.buttonStart.Text = "Start"
            '
            'labelLapPrompt
            '
            Me.labelLapPrompt.Font = New System.Drawing.Font("Comic Sans MS", 14.25!, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, CType(0, Byte))
            Me.labelLapPrompt.ForeColor = System.Drawing.Color.Yellow
            Me.labelLapPrompt.Location = New System.Drawing.Point(24, 87)
            Me.labelLapPrompt.Margin = New System.Windows.Forms.Padding(3, 1, 3, 3)
            Me.labelLapPrompt.Name = "labelLapPrompt"
            Me.labelLapPrompt.Size = New System.Drawing.Size(56, 25)
            Me.labelLapPrompt.TabIndex = 6
            Me.labelLapPrompt.Text = "Lap"
            Me.labelLapPrompt.Visible = False
            '
            'LabelTimePrompt
            '
            Me.LabelTimePrompt.Font = New System.Drawing.Font("Comic Sans MS", 15.75!, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, CType(0, Byte))
            Me.LabelTimePrompt.ForeColor = System.Drawing.Color.Yellow
            Me.LabelTimePrompt.Location = New System.Drawing.Point(16, 26)
            Me.LabelTimePrompt.Margin = New System.Windows.Forms.Padding(3, 0, 3, 3)
            Me.LabelTimePrompt.Name = "LabelTimePrompt"
            Me.LabelTimePrompt.Size = New System.Drawing.Size(88, 26)
            Me.LabelTimePrompt.TabIndex = 7
            Me.LabelTimePrompt.Text = "Time"
            '
            'labelBottomLeft
            '
            Me.labelBottomLeft.BackColor = System.Drawing.Color.Black
            Me.labelBottomLeft.ForeColor = System.Drawing.Color.Black
            Me.labelBottomLeft.Location = New System.Drawing.Point(8, 177)
            Me.labelBottomLeft.Margin = New System.Windows.Forms.Padding(3, 3, 1, 3)
            Me.labelBottomLeft.Name = "labelBottomLeft"
            Me.labelBottomLeft.Size = New System.Drawing.Size(16, 17)
            Me.labelBottomLeft.TabIndex = 8
            '
            'labelBottomRight
            '
            Me.labelBottomRight.BackColor = System.Drawing.Color.Black
            Me.labelBottomRight.ForeColor = System.Drawing.Color.Black
            Me.labelBottomRight.Location = New System.Drawing.Point(176, 177)
            Me.labelBottomRight.Margin = New System.Windows.Forms.Padding(1, 3, 3, 3)
            Me.labelBottomRight.Name = "labelBottomRight"
            Me.labelBottomRight.Size = New System.Drawing.Size(16, 17)
            Me.labelBottomRight.TabIndex = 9
            '
            'labelTopLeft
            '
            Me.labelTopLeft.BackColor = System.Drawing.Color.Black
            Me.labelTopLeft.ForeColor = System.Drawing.Color.Black
            Me.labelTopLeft.Location = New System.Drawing.Point(8, 8)
            Me.labelTopLeft.Margin = New System.Windows.Forms.Padding(3, 3, 3, 0)
            Me.labelTopLeft.Name = "labelTopLeft"
            Me.labelTopLeft.Size = New System.Drawing.Size(16, 17)
            Me.labelTopLeft.TabIndex = 10
            '
            'labelTopRight
            '
            Me.labelTopRight.BackColor = System.Drawing.Color.Black
            Me.labelTopRight.ForeColor = System.Drawing.Color.Black
            Me.labelTopRight.Location = New System.Drawing.Point(176, 8)
            Me.labelTopRight.Name = "labelTopRight"
            Me.labelTopRight.Size = New System.Drawing.Size(16, 17)
            Me.labelTopRight.TabIndex = 11
            '
            'FormMain
            '
            Me.AutoScaleDimensions = New System.Drawing.Size(9, 23)
            Me.BackColor = System.Drawing.Color.Navy
            Me.ClientSize = New System.Drawing.Size(200, 200)
            Me.Controls.Add(Me.labelTopRight)
            Me.Controls.Add(Me.labelTopLeft)
            Me.Controls.Add(Me.labelBottomRight)
            Me.Controls.Add(Me.labelBottomLeft)
            Me.Controls.Add(Me.LabelTimePrompt)
            Me.Controls.Add(Me.labelLapPrompt)
            Me.Controls.Add(Me.labelLap)
            Me.Controls.Add(Me.buttonLap)
            Me.Controls.Add(Me.buttonStart)
            Me.Controls.Add(Me.labelTime)
            Me.Controls.Add(Me.labelExit)
            Me.Font = New System.Drawing.Font("Microsoft Sans Serif", 14.25!, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, CType(0, Byte))
            Me.FormBorderStyle = System.Windows.Forms.FormBorderStyle.None
            Me.Name = "FormMain"
            Me.StartPosition = System.Windows.Forms.FormStartPosition.CenterScreen
            Me.Text = "StopWatch Demo"
            Me.TopMost = True
            Me.ResumeLayout(False)

        End Sub

#End Region

        Dim sw As New Stopwatch
        Dim paused As Boolean = False

        Private Sub formMain_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
            DrawForm()
        End Sub

        Private Sub DrawForm()
            'Shape the viewer form to look like a windshield
            Dim GraphicsPath As New Drawing2D.GraphicsPath

            GraphicsPath.AddArc(0, 0, 100, 100, 180, 90)
            GraphicsPath.AddArc(100, 0, 100, 100, 270, 90)
            GraphicsPath.AddArc(100, 100, 100, 100, 0, 90)
            GraphicsPath.AddArc(0, 100, 100, 100, 90, 90)

            Me.Region = New Region(GraphicsPath)
        End Sub

        Private Sub labelExit_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles labelExit.Click
            Me.Close()
        End Sub

        ' StopWatch code, easy as pie
        Private Sub buttonStart_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles buttonStart.Click
            If sw.IsRunning Then
                sw.Stop()
                buttonStart.Text = "Start"
                buttonLap.Text = "Reset"
            Else
                sw.Start()
                buttonStart.Text = "Stop"
                buttonLap.Text = "Lap"
                labelLap.Visible = False
                labelLapPrompt.Visible = False
            End If

        End Sub

        ' StopWatch code, easy as pie
        Private Sub buttonLap_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles buttonLap.Click
            If buttonLap.Text = "Lap" Then
                If sw.IsRunning Then
                    paused = True
                    labelLap.Visible = True
                    labelLapPrompt.Visible = True
                End If
            Else
                labelLap.Visible = False
                labelLapPrompt.Visible = False
                labelTime.Text = "00:00:00.00"
                buttonLap.Text = "Lap"
                sw.Reset()
            End If
        End Sub

        Private Sub timerMain_Tick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles timerMain.Tick
            If sw.IsRunning Then
                Dim ts As TimeSpan = TimeSpan.FromMilliseconds(sw.ElapsedMilliseconds)
                labelTime.Text = String.Format(CultureInfo.CurrentCulture, "{0:00}:{1:00}:{2:00}.{3:00}", ts.Hours, ts.Minutes, ts.Seconds, ts.Milliseconds / 10)

                If paused Then
                    labelLap.Text = labelTime.Text
                    labelLapPrompt.Text = "Lap"
                    paused = False
                End If
            End If
        End Sub

        Private Sub labelTopLeft_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles labelTopLeft.Click
            Me.Location = New Point(0, 0)
        End Sub

        Private Sub labelBottomLeft_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles labelBottomLeft.Click
            Me.Location = New Point(0, Screen.PrimaryScreen.WorkingArea.Height - Me.Height)
        End Sub

        Private Sub labelTopRight_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles labelTopRight.Click
            Me.Location = New Point(Screen.PrimaryScreen.WorkingArea.Width - Me.Width, 0)
        End Sub

        Private Sub labelBottomRight_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles labelBottomRight.Click
            Me.Location = New Point(Screen.PrimaryScreen.WorkingArea.Width - Me.Width, Screen.PrimaryScreen.WorkingArea.Height - Me.Height)
        End Sub

        Private Sub formMain_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Click
            Me.Location = New Point(CInt((Screen.PrimaryScreen.WorkingArea.Width - Me.Width) / 2), CInt((Screen.PrimaryScreen.WorkingArea.Height - Me.Height) / 2))
        End Sub

        Private Sub LabelTimePrompt_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles LabelTimePrompt.Click
            Me.Location = New Point(CInt((Screen.PrimaryScreen.WorkingArea.Width - Me.Width) / 2), CInt((Screen.PrimaryScreen.WorkingArea.Height - Me.Height) / 2))
        End Sub

        Private Sub labelLap_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles labelLap.Click
            Me.Location = New Point(CInt((Screen.PrimaryScreen.WorkingArea.Width - Me.Width) / 2), CInt((Screen.PrimaryScreen.WorkingArea.Height - Me.Height) / 2))
        End Sub

        Private Sub labelLapPrompt_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles labelLapPrompt.Click
            Me.Location = New Point(CInt((Screen.PrimaryScreen.WorkingArea.Width - Me.Width) / 2), CInt((Screen.PrimaryScreen.WorkingArea.Height - Me.Height) / 2))
        End Sub

        Private Sub labelTime_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles labelTime.Click
            Me.Location = New Point(CInt((Screen.PrimaryScreen.WorkingArea.Width - Me.Width) / 2), CInt((Screen.PrimaryScreen.WorkingArea.Height - Me.Height) / 2))
        End Sub
    End Class
End Namespace